modules/diskq/qdisk.c

```diff
@@ -561,38 +560,106 @@ qdisk_pop_head(QDisk *self, GString *record)
     return FALSE;
 
   if (self->hdr->read_head > self->hdr->write_head)
-    self->hdr->read_head = _correct_position_if_max_size_is_reached(self, &self->hdr->read_head);
+    self->hdr->read_head = _correct_position_if_max_size_is_reached(self, self->hdr->read_head);
 
   guint32 record_length;
-  if (!_try_reading_record_length(self, &record_length))
+  if (!_try_reading_record_length(self, self->hdr->read_head, &record_length))
     return FALSE;
 
   if (!_read_record_from_disk(self, record, record_length))
     return FALSE;
 
-  _update_positions_after_read(self, record_length);
+  _update_position_after_read(self, record_length, &self->hdr->read_head);
+  self->hdr->length--;
+  self->hdr->backlog_len++;
 
+  _maybe_apply_non_reliable_corrections(self);
+  return TRUE;
+}
+
+static gboolean
+qdisk_skip_record(QDisk *self, gint64 position, gint64 *new_position)
+{
+  if (position == self->hdr->write_head)
+    return FALSE;
+
+  if (position > self->hdr->write_head)
+    position = _correct_position_if_max_size_is_reached(self, position);
+
+  *new_position = position;
+
+  guint32 record_length;
+  if (!_try_reading_record_length(self, *new_position, &record_length))
+    return FALSE;
+
+  _update_position_after_read(self, record_length, new_position);
   return TRUE;
 }
 
 gboolean
 qdisk_remove_head(QDisk *self)
 {
-  if (self->hdr->read_head == self->hdr->write_head)
+  gboolean success = qdisk_skip_record(self, self->hdr->read_head, &self->hdr->read_head);
+
+  if (success)
+    {
+      self->hdr->length--;
+      self->hdr->backlog_len++;
+      _maybe_apply_non_reliable_corrections(self);
+    }
+
+  return success;
+}
+
+gboolean
+qdisk_ack_backlog(QDisk *self)
+{
+  if (self->hdr->backlog_len <= 0)
```

Comment:
Is it possible for backlog_len to be negative? If not, how about changing "<=" to "=="?

Reply:
Done, thanks.